website/blog/2021-02-04-raptorx.md

```diff
@@ -0,0 +1,152 @@
+![Presto Architecture](/img/blog/2021-02-04-raptorx/presto-arch.png)
+
+
+
+
+## RaptorX: Build a Hierarchical Caching Solution
+
+Historically, to solve the IO bound issue, Presto has a built-in Raptor connector to load data from remote storage to local SSD for fast access. However, this solution is no different from having a shared compute/storage node, which is against the idea of storage-compute disaggregation. The downside is obvious: either we waste CPU because the SSD in a worker is full or wasting SSD capacity if we are CPU bound. Thus, we started project RaptorX.
+
+
+
+RaptorX is an internal project name aiming to boost query performance for Presto by at least 10X. Hierarchical cache is part of the project. Cache is particularly useful when storage nodes are disaggregated from compute nodes. The goal for RaptorX is not to have a new connector or product, but a built-in solution so that existing workload can seamlessly benefit from it without migration. Especially, we target the existing Hive connector workload.
+
+
+
+The following figure shows the architecture of the caching solution. We have hierarchical layers of cache, which will be introduced in details in the remainder of this note:
+
+* Metastore versioned cache: We cache table/partition information in coordinator. Given metadata is mutable, like Iceberg or Delta Lake, the information is versioned. We only sync versions with metastore and fetch updated metadata when version is out of date.
+* File list cache: Cache file list from remote storage partition directory.
+* Fragment result cache: Cache partially computed result on leaf workers’ local SSDs. Pruning techniques are required to simplify query plans as they change all the time.
+* File handle and footer cache: Cache open file descriptor and stripe/file footer information in leaf worker memory. These pieces of data are mostly frequently accessed when reading files.
+* Alluxio data cache: Cache file segments with 1MB aligned chunks on leaf workers’ local SSDs. The library is built from Alluxio cache service.
+* Affinity scheduler: A scheduler that sends sticky requests based on file paths to particular workers to maximize cache hit rate.
+
+
+
+![RaptorX Architecture](/img/blog/2021-02-04-raptorx/raptorx-arch.png)
+
+
+## Metastore Versioned Cache
+
+A Presto coordinator caches table metadata (schema, partition list, and partition info) to avoid long `getPartitions` calls to Hive Metastore. However, Hive table metadata is mutable. Versioning is needed to determine if the cached metadata is valid or not. To achieve that, the coordinator attaches a version number to each cache key-value pair. When a read request comes, the coordinator asks the Hive Metastore to get the partition info (if it's not cached at all) or confirm with Hive Metastore to check if cached info is up to date. Though the roundtrip to Hive Metastore cannot be avoided, the version matching is relatively cheap compared with fetching the entire partition info.
```

Comment:
```suggestion
A Presto coordinator caches table metadata (schema, partition list, and partition info) to avoid long `getPartitions` calls to Hive Metastore. However, Hive table metadata is mutable. Versioning is needed to determine if the cached metadata is valid or not. To achieve that, the coordinator attaches a version number to each cache key-value pair. When a read request comes, the coordinator asks the Hive Metastore to get the partition info (if it's not cached at all) or checks with Hive Metastore to confirm the cached info is up to date. Though the roundtrip to Hive Metastore cannot be avoided, the version matching is relatively cheap compared with fetching the entire partition info.
```